### DISCUSSION TITLE
What is algorothm of Get-Random?

### DISCUSSION THREAD

**Discussion by 237dmitry** [PowerShell/PowerShell#20940] [General]
I have music collection:

```powershell
$ (dir ~/Music/*.flac -s).Count
4786
```

Daily I play randomly 20-50 tracks. Over the past four years, since I started keeping statistics on the count of plays and ratings of each track, I have been perplexed. Some tracks are chosen much more often than others. Why, out of the total number of tracks, has one already been played 16 times, and the rest 1-5 times?

Most popular (and only added to the collection in February 2023):

```powershell
$ metaflac --show-all './Music/W.A.S.P/2004. The Neon God - Part 2 - The Demise/01 - Never Say Die.flac'
ADDED_TIMESTAMP=133213879373379659            # 2023/02/20 20:32:17
ALBUM=The Neon God - Part 2 - The Demise
ALBUMARTIST=W.A.S.P.
ARTIST=W.A.S.P.
DATE=2004
FIRST_PLAYED_TIMESTAMP=133217375711599289    # 2023/02/24 21:39:31
GENRE=Heavy Metal
RATING=5
REPLAYGAIN_ALBUM_GAIN=-7.40 dB
REPLAYGAIN_ALBUM_PEAK=0.944092
REPLAYGAIN_TRACK_GAIN=-7.99 dB
REPLAYGAIN_TRACK_PEAK=0.943359
TITLE=Never Say Die
TRACKNUMBER=01
LAST_PLAYED_TIMESTAMP=133463284622591493     # 2023/12/06 12:27:42
PLAY_COUNT=16
```

**Comment by ThomasNieto** [comment]
```
/// <summary>
/// Provides an adapter API for random numbers that may be either cryptographically random, or
/// generated with the regular pseudo-random number generator. Re-implementations of
/// methods using the NextBytes() primitive based on the CLR implementation:
///     https://referencesource.microsoft.com/#mscorlib/system/random.cs.
/// </summary>
```

https://github.com/PowerShell/PowerShell/blob/103c945c325328ac37d441488bcd6a38adf43888/src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetRandomCommandBase.cs#L558C5-L563C4

**Reply by 237dmitry** [reply]
Sorry, but my question was rhetorical. I don't understand the pattern of choosing a random item. Given that this choice is repeated much more often than other items. I have only one explanation: the Get-Random cmdlet loves WASP. Can the selection from 4,500 variants of the same element 16 times a year be considered logical, while there are elements that were selected only once in three years? Or is it just a coincidence?

**Reply by SeeminglyScience** [reply]
RNG is RNG, it can be surprising sometimes for sure. If you can find something reproducible, we can look into it, but the sample size here is small enough to call it normal RNG

**Reply by 237dmitry** [reply]
> If you can find something reproducible, we can look into it
 
I can't. This is too long period of time and many attempts of random choosing in order to reproduce (or not) my statistic. I believe that it just happened that way.

**Comment by rhubarb-geek-nz** [comment]
> I don't understand the pattern of choosing a random item.

In theory there should be no pattern if it is a true random selection.

**Reply by 237dmitry** [reply]
> In theory there should be no pattern if it is a true random selection.

The selection is simple:

```powershell
$Global:playlist = [Collections.Queue]::new()
$files = Get-Childitem $Directory -Include *.flac -Recurse | Get-Random -Count 10
foreach ($file in $files) { $Global:playlist.Enqueue($file) }
```

**Reply by rhubarb-geek-nz** [reply]
What I mean is there should be no pattern to the sequence of random numbers you get, there should be no "sequence" except just one number after the other. There should be no guarantee that over a period of time each value should have an even distribution. While you might expect 50% heads and 50% tails, a true random generator makes no such guarantee, each random number should be independent from the previous value and not time dependent.

**Reply by 237dmitry** [reply]
I have no explanation why the element with a conditional (since it is not a constant value) index 3882 is selected much more often than others.

```powershell
$ & {
         $files = dir ~/Music/*.flac -s
         $item = $files | ? { $_.Name -eq '01 - Never Say Die.flac' }
         [array]::IndexOf($files, $item)
    }
3832

```

**Comment by iSazonov** [comment]
`Get-Random` without `Seed` parameter and `Get-SecureRandom` use default implementation [RandomNumberGenerator](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.randomnumbergenerator?view=net-8.0). It is _cryptographically_ strong random values. Its implementation depends on OS.

`Get-Random` with `Seed` parameter use [Random class](https://learn.microsoft.com/en-us/dotnet/api/system.random?view=net-8.0). It is a [pseudo-random number generator](https://learn.microsoft.com/en-us/dotnet/api/system.random?view=net-8.0#remarks).

**Reply by 237dmitry** [reply]
In my case, there is no need to use seed. I opened this discussion only because I was very surprised by the playback statistics. Considering that these statistics cover a significant period of time, an uncountable number of samples independent of each other, there is no point in looking for technical or algorithmic patterns in the selection of elements, this is simply surprising. The music collection, as the basis of choice, is not static, it changes, and the hierarchy of files changes accordingly. I view this as an oft-repeated random choice. At some point I even thought that Get-Random had musical preferences. Or somewhere under the hood it analyzes the properties of the files, which, of course, is not the case.
